tt_eager/tt_dnn/op_library/sdpa/kernels/dataflow/reader_decode_interleaved.cpp

```diff
@@ -0,0 +1,211 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+#include "dataflow_api.h"
+
+#include "debug/dprint.h"
+
+template<uint32_t tile_bytes, uint32_t num_readers>
+constexpr uint32_t get_barrier_read_threshold() {
+     return ((512 / num_readers) * (1024 + 128)) / tile_bytes;
+ }
+
+void kernel_main() {
+    /*
+    In DRAM, Q is (B, PNHt, DHt), K is (B, St, DHt), V is (B, St, DHt), mask is (B, PNHt, PSt)
+    We want to read for a particular batch cur_batch, and sequence length up to padded layer length.
+    We read Q: (cur_batch, PNHt, DHt), K: (cur_batch, PSt, DHt), V: (cur_batch, PSt, DHt), mask: (cur_batch, PNHt, PSt)
+    */
+    constexpr uint32_t B = get_compile_time_arg_val(0);  // batch size
+    constexpr uint32_t PNHt = get_compile_time_arg_val(1);  // padded number of heads in tiles
+    constexpr uint32_t PSt = get_compile_time_arg_val(2);  // padded layer length in tiles
+    constexpr uint32_t St = get_compile_time_arg_val(3);  // full sequence length of kv cache in tiles
+    constexpr uint32_t DHt = get_compile_time_arg_val(4);  // head dim
+    constexpr uint32_t Sk_chunk_t = get_compile_time_arg_val(5);  // number of tiles in seqlen of a k/v/mask chunk
+    constexpr uint32_t k_num_chunks = get_compile_time_arg_val(6);  // number of chunks in K, where k_num_chunks*Sk_chunk_t = PSt
+    constexpr uint32_t num_cores = get_compile_time_arg_val(7);
+    constexpr uint32_t cur_batch =  get_compile_time_arg_val(8);
+    constexpr uint32_t k_chunk_start = get_compile_time_arg_val(9);
+    constexpr uint32_t k_chunk_end = get_compile_time_arg_val(10);
+    constexpr bool is_q_sharded = get_compile_time_arg_val(11);
```

Comment:
rename this file since input could be interleaved or sharded?

Reply:
I kept it reader_decoder_interleaved because most inputs (e.g. k, v, and mask) are interleaved. Q is optional interleaved or sharded

Reply:
I changed it to reader_decoder_all.cpp